### DISCUSSION TITLE
How to migrate from older versions?

### DISCUSSION THREAD

**Discussion by e-Evolution** [adempiere/adempiere#4103] [Announcements]
Steps to migrate between ADempiere versions

1.- Make a new installation of Adempiere with the new binary Adempiere_394LTS.tar.gz on a test server
2.- Import a backup of the production version
3.- Make configuration of the new environment RUN_Setup
4.- Execute the migration of utils/RUN_MigrateXML.sh force parameters and validate that everything is fine, if there are problems you must resolve any process during the migration.
`./RUN_MigrateXML.sh force`

When executing the forced migration process, it forces all the scripts to be applied even if one fails, an error can affect other XML because there are dependencies, the advantage of this is that you can later review which steps failed and resolve each case until all the steps are applied ending with a successful migration

5.- Validate if your customizations are compatible, if it is not required to migrate the code, the views and modifications of the application dictionary
6.- The client to whom the migration was carried out has a trial period if this does not present inconveniences the migration is released for production
7- Execute from step 1 to 6 on the production server and without active users.

Any question is welcome

**Comment by yuanaw8076** [comment]
Dear Victor

The meaning of "2.- Import a backup of the production version" is the Database from previous version?

Thanks

**Comment by PraponVDevSoft** [comment]
thank you Victor

**Comment by marcalwestf** [comment]
## Migrate From ADempiere Seed Database To Target Database
Under certain circumstances, it is better to migrate directly from one database to the target database.
Directly means here,  without XML import files.

This can be the case when for example 
- there are many customizations in the target database which collide with the XMLs or 
- the legacy version is too old and there would be too many XMLs to be executed.

### How it works
A _Migrate_ ADempiere Process migrates from seed to target at  Postgres Data Dictionary level.

### Configuration Needed
The configuration is simple: 
- a **seed database** is needed e.g. ADempiere 394 and also 
- a **target database** e.g. a legacy database from 3.9.2.

Both databases must exist on the database server.
The result will be a target database in sync with the lastest seed.

### Start Process
There are two ways to start the Migration process:
- With a script
Execute script _$ADEMPIERE_HOME/utils/RUN_Migrate.sh_ from console.
Follow the instructions from the next section.
- From _Eclipse/IntelliJ_
Create a project, taking the classs _com.kkalice.adempiere.migrate.Migrate_ as the main (=executable) class.
Follow the instructions from the next section.
![03-Migration_01](https://user-images.githubusercontent.com/1789408/236709975-12fa7fa5-fd0b-46e1-ac8e-7d99afdc11c7.png)

### Execute the process
Migration Tool initial window:
![04-Migration_02](https://user-images.githubusercontent.com/1789408/236710063-c80c524b-bad3-4160-805d-0690952aa29e.png)

Select source (=seed) and target (=legacy) database:
![05-Migration_03](https://user-images.githubusercontent.com/1789408/236710117-a5fe99e0-45b7-4e0e-b590-48a7423755f5.png)

Hints:

- It is far better to exectue the migration from _Eclipse/IntelliJ_
- Make sure to configure the IDE in a way that the process only displays warning logs.
Otherwise, and depending on the database size, a **heap error** may be thrown.
- Execute the process with **breakpoints**. 
This way you will know if a step was done correctly and you can proceed to the next one.
- The program displays errors and warnings.
Also, the program shows where the log files are situated, e.g. _migration_YYYYMMDDHHMM.trace.log_, _migration_YYYYMMDDHHMM.error.log_.
- You can mostly ignore the warnings.
- The errors must be corrected, starting with te first.
- Often, the first error causes the next errors to be thrown, so it is advisable to correct the first error and start anew.
- The error gives you a hint on the error. 
Correct the error on the **initial target database** (this means, discard the results of the migration process), make a backup (in case the bug fix is wrong) and start the migration anew. If your correction was OK, the error will disappear.
- Do the iteration of error display-correction on initial target-backup of initial target-restart of migration tool as long as there is no error displayed.
- The result will be a target database in sync with the lastest seed.
- This may take days to execute, depending on the complexity and the changes of the legacy database.


### Links
See 
- http://wiki.adempiere.net/Universal_Migration_Tool
- http://wiki.adempiere.net/Migrate_-_Introduction#N10137
- http://wiki.adempiere.net/Migrate_-_Migrating_a_Database#N104AA
